### ISSUE TITLE
TODO_1: Getting Started with the Code

### ISSUE THREAD

**Issue by yassersouri** [yassersouri/edges#1]
Here are the list of things that needs to be doen before I can say that I know how the code works and can imporve upon it:

* [ ] Get the training data and train a new model.
* [ ] Test the new model on single images.
* [ ] Evaluate the model on BSDS500 and arrive at the same results as reported in the papers.
* [ ] See how the `edgesChns` function prepares the features and how can we define our own set of features, instead. @erfannoury Do you know what are `chnsReg` and `chnsSim`, the outputs of `edgesChns`? Did you see anything about _self-similarity channels_ in the paper? I suggest you take a look at the comments inside the `edgesChns` file first.
* [ ] Try to make a change to the features and see how it affects the results.

**Comment by yassersouri** [comment]
Early concern: Models learnt on hypercolumn features need a lot of training data. Since the number training images in BSDS500 is 200 (which is very small) maybe we get stuck here!

A possible solution that comes to mind is to change the parameters of the code such as `nPos` and `nNeg` while training.

**Comment by erfannoury** [comment]
So you have done the training. I will try to do so tonight. I think evaluation would be straight-forward. 
And about the `edgesChns`, I will take a look asap.

**Comment by erfannoury** [comment]
Re early concern: Since random forest is trained on patches of size `2 x 2` I don't think we would be short of training data, however more training data won't hurt.

**Comment by yassersouri** [comment]
I think @pdollar has done a great job writing this software. It is very easy to train and test new models.

Training the model on BSDS500 again, I get the following results:
ODS=0.739 OIS=0.758 AP=0.791 R50=0.927

They seem to match the results on the ICCV paper. @erfannoury Any ideas?

**Comment by yassersouri** [comment]
I will try training a different model with anothre randomness seed to get an idea of how random is the whole process.

**Comment by erfannoury** [comment]
Yes, it seems to be correct. I have found randomness factor to be so import in training RandomForests.

**Comment by yassersouri** [comment]
I have tried it two more times with different random seeds:

ODS=0.738 OIS=0.758 AP=0.790 R50=0.926

ODS=0.738 OIS=0.757 AP=0.791 R50=0.926

It seems to be okay in regards to the effect of randomness.

**Comment by erfannoury** [comment]
Yes, results shows consistency against the random factor.

**Comment by yassersouri** [comment]
##Regarding Features:
I investigated the code in side `edgesChns.m` carefully. With the default parametrs it seems that the only difference between `chnsReg` and `chnsSim` is the smoothing kernel size applied to them at the end.

One issue is that their feature dimension is 13! Whereas ours will be in O(10^3) dimensions! I think this will cause problems with our limited (32Gb) RAM.

For the sake of completeness read my comments in the recent commit 11435c8cf12.

**Comment by erfannoury** [comment]
So its features are low-dimensional. But ours is very high-dimensional and memory issue is a highly serious issue, believe me! :smiley: 
What do you think? What can we do? Can't we change the classifier to something else, something batch-wise or SGD-trainable?

**Comment by pdollar** [comment]
Hi -
Just wanted to let you know I'm cc-ed on all of these discussions, not sure
if that was your intention.
Thanks,
Piotr

**Comment by erfannoury** [comment]
Hello Dr. Dollar, 
We are sorry for this inconvenience. We didn't know about this issue. 
We make sure this won't happen again.

Best Regards, 
Erfan Noury

**Comment by pdollar** [comment]
No problem :)
Good luck w your research!
Piotr

**Comment by yassersouri** [comment]
Sorry. That was because I mentioned you in one of the comments.

Please unsubscribe to this thread if it is bothering you. 

Thanks for the amazing code!

**Comment by yassersouri** [comment]
BTW since we have your attention let me ask a question.

In `edgesChns.m`, what is the difference between the two output parameters (
chnsReg and chnsSim). As of now I think the only difference is the size of
the smoothing kernel applied on them at the end. But the names suggest
otherwise. Can you comment?

P.S: I noticed that you might have unsubscribed from the thread so I cced
your email. Sorry for inconveniences.

**Comment by pdollar** [comment]
The outputs are:
% OUTPUTS
%  chnsReg    - [h x w x nChannel] regular output channels
%  chnsSim    - [h x w x nChannel] self-similarity output channels
See the paper for details.

Cheers,
  -Piotr
